File: kotlin/src/main/kotlin/ch/frostnova/example/shoppingcart/model/Entity.kt
package ch.frostnova.example.shoppingcart.model

abstract class Entity<ID> {

    var id: ID? = null

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Entity<*>

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }
}

File: typescript/src/model/entity.ts
export abstract class Entity<T> {

    constructor(private type: string, public id?: T){}

    public equals(obj: Entity<T>): boolean {
        if (obj === this) {
            return true;
        }
        if (obj == null || obj.type !== this.type) {
            return false;
        }
        return (obj as Entity<T>).id === this.id;
    }
}